src/midgard/point_tile_index.cc

```diff
@@ -0,0 +1,100 @@
+#include "midgard/point_tile_index.h"
+
+namespace valhalla {
+
+namespace midgard {
+
+// A "special" value that means "this point is deleted". An invalid
+// lat/lon.
+const PointLL PointTileIndex::deleted_point = {1000.0, 1000.0};
+
+PointTileIndex::PointTileIndex(double tile_width_degrees) {
+  int level = 1;
+  num_x_subdivisions = 1;
+  double degrees_at_level = 180.0;
+  constexpr int max_level = 32;
+  while ((degrees_at_level > tile_width_degrees) && (level <= max_level)) {
+    level++;
+    degrees_at_level /= 2.0;
+    num_x_subdivisions = num_x_subdivisions << 1;
+  }
+
+  // y's space is half that of x (because the full latitude range is
+  // 180 deg vs longitude which is 360).
+  num_y_subdivisions = num_x_subdivisions >> 1;
+}
+
+template <class container_t> void PointTileIndex::tile(const container_t& polyline) {
+  this->points.reserve(polyline.size());
+  tiled_space.reserve(polyline.size());
+  size_t index = 0;
+  for (auto iter = polyline.begin(); iter != polyline.end(); iter++, index++) {
+    this->points.emplace_back(*iter);
+    TileId pid = get_tile_id(*iter);
+    auto map_iter = tiled_space.find(pid);
+    if (map_iter == tiled_space.end())
+      tiled_space.insert(std::pair<TileId, std::unordered_set<size_t>>{pid, {index}});
+    else
+      map_iter->second.insert(index);
+  }
+}
+
+void PointTileIndex::get_points_near(const PointLL& pt, std::unordered_set<size_t>& near_pts) {
+  TileId pid = get_tile_id(pt);
+  uint32_t px = prevx(pid.x);
+  uint32_t nx = nextx(nextx(pid.x));
+  uint32_t py = prevy(pid.y);
+  uint32_t ny = nexty(nexty(pid.y));
+  for (uint32_t x = px; x != nx; x = nextx(x)) {
+    for (uint32_t y = py; y != ny; y = nexty(y)) {
+      auto iter = tiled_space.find(TileId{x, y});
+      if (iter != tiled_space.end()) {
+        near_pts.insert(iter->second.begin(), iter->second.end());
+      }
+    }
+  }
+}
+
+void PointTileIndex::get_points_near_segment(const LineSegment2<PointLL>& seg,
+                                             std::unordered_set<size_t>& near_pts) {
```

Comment:
why dont we just return near points
```suggestion
std::unordered_set<size_t> PointTileIndex::get_points_near(const PointLL& pt)
std::unordered_set<size_t> PointTileIndex::get_points_near_segment(const LineSegment2<PointLL>& seg)
```

Reply:
Yea I think about that a lot.  C++ has evolved a lot and is making every effort (via optimizations like RVO, copy elison, move operators) to make returning complex objects an efficient and idiomatic part of the language.  I haven't checked lately, but some of those optimizations I mentioned are disabled in debug builds.  Hence, when returning complex objects, the debug build will fallback to copy-construction which would suffer performance.  Also, I've read that it is easy to "defeat" RVO by simply adding early `return`s to your function.  

I certainly have some more learning to do on these subjects, especially since compilers are making continuous improvements on these fronts.  However, I can say with (almost) absolute certainty that the manner in which I've written those methods is as efficient as can be written.  Nothing is left to hoping the compiler optimizer moves things efficiently - and debug builds suffer nothing.

All that said, I'm certainly willing to modify those routines as suggested.  Being correct is priority zero - being consistent is right behind.  Just wanted to share my thinking.  thanks.